Orchestrator state_store: ENOENT race in _ensure_worktree zombies pipelines (status stays awaiting_human, no FAILED transition)

## Symptom

A live pipeline (`issue-1557`, refine phase, `awaiting_human` with `decision-1` pending) had its `_run_pipeline` execution loop die mid-flight. The orchestrator was unable to mark the pipeline FAILED, so the status stayed `awaiting_human` forever. A subsequent host approval (`provide_input` resolving `decision-1`) was accepted by the decision queue at 02:38:42 but no one consumed it — the pipeline runner had died ~8 minutes earlier at 02:30:07. From a host's point of view the pipeline looked alive (status=`awaiting_human`, `wait-status` correctly silent) but was actually a zombie.

## Root cause: ENOENT race in `_ensure_worktree`

`orchestrator/state_store.py::_ensure_worktree` (lines 242–279):

```python
wt = self._worktree_dir

if wt.exists():                                      # line 244 — TOCTOU check
    healthy = False
    for _attempt in range(2):
        result = self._run_git("rev-parse", "--is-inside-work-tree", cwd=wt, check=False)
        ...
    if healthy:
        return wt

    logger.warning("Worktree validation failed, recreating: ...")
    try:
        shutil.rmtree(wt)                            # line 275 — TOCTOU use
    except OSError as exc:
        raise GitOperationError(
            f"Failed to remove stale state worktree at {wt}: {exc}"
        ) from exc
```

Between the `wt.exists()` probe at line 244 and the `shutil.rmtree(wt)` at line 275, another caller deleted the worktree directory. `shutil.rmtree` then raises `FileNotFoundError` — which is *exactly the desired post-state*: the directory is gone, we wanted it gone, that's the win condition. But the catch-all `except OSError` reraises `FileNotFoundError` as a `GitOperationError`, killing the pipeline runner.

In this incident the racing deleter was the orchestrator's own pipeline cleanup path — at 02:30:07 the same orchestrator logged `Pipeline worktrees cleaned up pipeline_id=issue-1557` (`kubernetes_spawner.py:951`) for several worktree IDs including the state worktree. The respawn flow for the overseer (`Overseer exited mid-pipeline, respawning` at 02:29:42) appears to have raced with the next phase's `_ensure_worktree` call.

**Fix**: treat ENOENT on `shutil.rmtree` as success. Either pass `ignore_errors=True` and re-check `wt.exists()`, or:

```python
try:
    shutil.rmtree(wt)
except FileNotFoundError:
    pass  # Already gone — that's what we wanted
except OSError as exc:
    raise GitOperationError(...) from exc
```

The post-rmtree code (lines 280+) already handles the wt-gone case (it calls `_remove_stale_admin_dir(force=True)` and only fails if `wt.exists()` after).

## Compounding bug: `mark_pipeline_failed` also crashes via the same path

When `_run_pipeline` catches the `GitOperationError` and tries to mark the pipeline FAILED, the FAILED-transition code path also calls `load_pipeline` → `pipelines_dir` → `worktree` → `_ensure_worktree` again. By the second call the worktree directory truly is gone, so the `wt.exists()` branch is skipped, but the next branch (line 301: `_state_branch_exists()` → `_add_worktree_with_branch_recovery(wt)`) fails with:

```
Git command failed: Preparing worktree (checking out 'egg/pipeline-state') failed
```

The mark-FAILED handler logs:

```
[ERROR] orchestrator.pipelines: Failed to mark pipeline FAILED after exception
        pipeline_id=issue-1557
        original_error="Failed to remove stale state worktree at ..."
        mark_error="Git command failed: Preparing worktree (checking out 'egg/pipeline-state') fa..."
        [/app/routes/pipelines.py:14644]
```

…and exits without setting `status=FAILED`. The pipeline JSON on disk continues to report `awaiting_human` and the host's `wait-status` correctly emits no events because no events occur — there's nothing alive to emit them.

**Fix**: the FAILED-transition path needs its own resilience. Options:
1. Make `_ensure_worktree` self-healing (recover from the `git worktree add` failure here too — the branch likely already exists locally; `--force` or a fallback to checkout-only is plausible).
2. Have `mark_pipeline_failed` write directly to the pipeline JSON via a path that does **not** require an active state worktree.
3. At minimum, on `mark_error`, surface the failure as a one-shot OVERSEER_ALERT or log a separate `pipeline.zombied` event so hosts have a signal that the runner is dead.

Without this, **any `_ensure_worktree` failure on the FAILED path silently zombies the pipeline.**

## Reproduction

The race is non-deterministic (depends on the cleanup goroutine timing), but the conditions in this incident were:
- Concurrent agents completing (refine phase, 3 agents — refiner + 2 reviewers).
- Overseer "Overseer exited mid-pipeline, respawning" cycle at 02:29:42 — likely the trigger for the cleanup that raced the next `_ensure_worktree` call.
- A pending HITL decision (`decision-1`, refine phase_gate) created at 02:28:04 — by 02:30:07 the runner was waiting in `decision_queue.wait_for_decision` and that's where the worktree access happened.

## Full traceback

```
2026-04-29 02:30:06 [ERROR] orchestrator.pipelines: Pipeline execution failed pipeline_id=issue-1557
                                error="Failed to remove stale state worktree at /home/egg/.egg-state/pipeline-worktr..."
Traceback (most recent call last):
  File "/app/state_store.py", line 275, in _ensure_worktree
    shutil.rmtree(wt)
  File "/usr/local/lib/python3.13/shutil.py", line 763, in rmtree
    _rmtree_safe_fd(stack, onexc)
  File "/usr/local/lib/python3.13/shutil.py", line 704, in _rmtree_safe_fd
    onexc(func, path, err)
  File "/usr/local/lib/python3.13/shutil.py", line 665, in _rmtree_safe_fd
    orig_st = os.lstat(name, dir_fd=dirfd)
FileNotFoundError: [Errno 2] No such file or directory:
                   PosixPath('/home/egg/.egg-state/pipeline-worktree-egg')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/app/routes/pipelines.py", line 14128, in _run_pipeline
    dq.wait_for_decision(decision.id)
  File "/app/decision_queue.py", line 317, in wait_for_decision
    decision = self.get_decision(decision_id)
  File "/app/decision_queue.py", line 207, in get_decision
    pipeline = self._load_pipeline()
  File "/app/decision_queue.py", line 81, in _load_pipeline
    return store.load_pipeline(self.pipeline_id)
  File "/app/state_store.py", line 590, in load_pipeline
    path = self._get_pipeline_path(pipeline_id)
  File "/app/state_store.py", line 472, in _get_pipeline_path
    path = self.pipelines_dir / f"{pipeline_id}.json"
  File "/app/state_store.py", line 206, in pipelines_dir
    return self.worktree / self.PIPELINES_DIR
  File "/app/state_store.py", line 201, in worktree
    self._worktree = self._ensure_worktree()
  File "/app/state_store.py", line 277, in _ensure_worktree
    raise GitOperationError(...) from exc
state_store.GitOperationError: Failed to remove stale state worktree at
    /home/egg/.egg-state/pipeline-worktree-egg:
    [Errno 2] No such file or directory: PosixPath('/home/egg/.egg-state/pipeline-worktree-egg')
```

The mark-FAILED path then crashes with `Git command failed: Preparing worktree (checking out 'egg/pipeline-state') fa...` (truncated), at `pipelines.py:14644`.

## Operator impact

This was the second silent-zombie pipeline this session — it presents identically to a healthy `awaiting_human` pipeline from the host's `/sdlc` skill perspective. `wait-status` correctly emits no events (nothing is happening), so the host-side liveness story works as designed; but the operator has no signal that the runner is dead until they go look at orchestrator logs. Memory `feedback_brc_silence_normal.md` correctly tells me silence is normal — and indeed it was correct here, just for the wrong reason.

## Suggested scope for the fix

1. **State_store ENOENT-on-rmtree**: small, targeted, one-line-style fix at `state_store.py:275`. Catch `FileNotFoundError` separately and treat as success; let other `OSError` subclasses keep raising.
2. **mark-FAILED resilience**: the FAILED path must succeed even when the state worktree is unrecoverable. Either fall back to a worktree-free write of the pipeline JSON (writing directly to the bare repo via `git update-ref` + a tree object), or add an in-memory `pipeline.status = FAILED` sentinel that the API surfaces even when persistence fails.
3. **Zombie detection signal**: emit a `pipeline.zombied` event (or an `OVERSEER_ALERT` of distinct type) when the FAILED handler can't persist, so hosts can break out of `wait-status` instead of silently waiting forever.

(1) on its own would have prevented this incident. (2) and (3) are belt-and-suspenders for the next state-store failure mode.

## References

- `orchestrator/state_store.py:275` — the raising rmtree
- `orchestrator/state_store.py:208–326` — full `_ensure_worktree`
- `orchestrator/routes/pipelines.py:14128` — `_run_pipeline` decision wait that surfaced the error
- `orchestrator/routes/pipelines.py:14611, :14644` — pipeline-failed and mark-failed-failed log sites
- `orchestrator/decision_queue.py:81, :207, :317` — chain that re-enters the worktree on every decision read
- The pipeline that hit this: `issue-1557` (already cancelled by the operator after this issue was filed)

--- a/orchestrator/routes/pipelines.py
+++ b/orchestrator/routes/pipelines.py
@@ -14648,6 +14648,32 @@ def _run_pipeline(
                 mark_error=str(fail_err),
                 exc_info=True,
             )
+            # Surface a synthetic ``pipeline.failed`` to the EventBus even
+            # though persistence failed.  Without this, hosts blocked on
+            # ``/status/wait`` (whose event allowlist requires
+            # pipeline.failed/completed/cancelled) wait forever on a dead
+            # runner — the zombie symptom in #2234.  The persisted JSON
+            # may still report ``running``/``awaiting_human``, but the
+            # event lets the host break out and surface the wedge to the
+            # operator.
+            if _emit_event is not None:
+                try:
+                    _emit_event(
+                        EventType.PIPELINE_FAILED,
+                        pipeline_id,
+                        data={
+                            "status": PipelineStatus.FAILED.value,
+                            "persisted": False,
+                            "original_error": str(e),
+                            "mark_error": str(fail_err),
+                        },
+                    )
+                except Exception as emit_err:
+                    logger.warning(
+                        "Failed to emit synthetic pipeline.failed event",
+                        pipeline_id=pipeline_id,
+                        error=str(emit_err),
+                    )
     finally:
         # Stop health monitor polling and unsubscribe from events
         if health_monitor_timer is not None:

--- a/orchestrator/routes/pipelines.py
+++ b/orchestrator/routes/pipelines.py
@@ -14611,6 +14611,7 @@ def _run_pipeline(
         logger.error(
             "Pipeline execution failed", pipeline_id=pipeline_id, error=str(e), exc_info=True
         )
+        persisted_ok = False
         try:
             store = get_state_store(repo_path)
             with get_pipeline_state_lock(pipeline_id):
@@ -14627,6 +14628,7 @@ def _run_pipeline(
                     pipeline.status = PipelineStatus.FAILED
                     pipeline.error = str(e)
                     store.save_pipeline(pipeline)
+                    persisted_ok = True
 
                 # Report pipeline failure to collaborator
                 report_pipeline_status(
@@ -14649,13 +14651,14 @@ def _run_pipeline(
                 exc_info=True,
             )
             # Surface a synthetic ``pipeline.failed`` to the EventBus even
-            # though persistence failed.  Without this, hosts blocked on
-            # ``/status/wait`` (whose event allowlist requires
+            # though the mark-FAILED block raised.  Without this, hosts
+            # blocked on ``/status/wait`` (whose event allowlist requires
             # pipeline.failed/completed/cancelled) wait forever on a dead
-            # runner — the zombie symptom in #2234.  The persisted JSON
-            # may still report ``running``/``awaiting_human``, but the
-            # event lets the host break out and surface the wedge to the
-            # operator.
+            # runner — the zombie symptom in #2234.  ``persisted`` carries
+            # whether ``save_pipeline`` actually flushed FAILED to disk
+            # before the inner block raised: True means disk state matches
+            # the event, False means consumers should treat the event as
+            # the only authoritative source.
             if _emit_event is not None:
                 try:
                     _emit_event(
@@ -14663,7 +14666,7 @@ def _run_pipeline(
                         pipeline_id,
                         data={
                             "status": PipelineStatus.FAILED.value,
-                            "persisted": False,
+                            "persisted": persisted_ok,
                             "original_error": str(e),
                             "mark_error": str(fail_err),
                         },

--- a/orchestrator/state_store.py
+++ b/orchestrator/state_store.py
@@ -273,6 +273,17 @@ class StateStore:
             )
             try:
                 shutil.rmtree(wt)
+            except FileNotFoundError:
+                # Concurrent caller (e.g. the state-store probe at
+                # ``state_store_probe.py``, a sibling pipeline thread, or
+                # the ``/api/v1/health`` probe) already cleaned up the
+                # worktree between our ``wt.exists()`` check and this
+                # rmtree.  ``_ensure_worktree`` is invoked from many
+                # threads concurrently and is not wrapped in
+                # ``_git_op()``, so the TOCTOU window is real.  ENOENT
+                # here is exactly the post-state we wanted — fall
+                # through to recreate.  Issue #2234.
+                pass
             except OSError as exc:
                 raise GitOperationError(
                     f"Failed to remove stale state worktree at {wt}: {exc}"

--- a/orchestrator/tests/test_pipeline_failure_path.py
+++ b/orchestrator/tests/test_pipeline_failure_path.py
@@ -170,6 +170,117 @@ class TestFailurePathEmitsPipelineEvent:
         )
 
 
+class TestZombiePipelineSyntheticEvent:
+    """Regression tests for #2234.
+
+    When ``_run_pipeline``'s catch-all exception handler tries to mark
+    the pipeline FAILED but that itself crashes (e.g., the racing
+    ``_ensure_worktree`` ENOENT, lock timeout, or any other
+    ``StateStoreError``), the host blocked on ``/status/wait`` would
+    wait forever — its event allowlist requires
+    ``pipeline.failed/completed/cancelled`` to unblock and no event
+    was emitted.  The fix surfaces a synthetic ``PIPELINE_FAILED``
+    event to the EventBus so the host wakes up on the wedge.
+    """
+
+    @patch("routes.pipelines._run_concurrent_phase")
+    @patch("routes.pipelines._emit_event")
+    @patch(_COMMON_PATCHES[7])
+    @patch(_COMMON_PATCHES[6])
+    @patch(_COMMON_PATCHES[5])
+    @patch(_COMMON_PATCHES[4])
+    @patch(_COMMON_PATCHES[3])
+    @patch(_COMMON_PATCHES[2])
+    @patch(_COMMON_PATCHES[1])
+    @patch(_COMMON_PATCHES[0])
+    def test_synthetic_pipeline_failed_event_when_mark_failed_crashes(
+        self,
+        mock_emit,
+        mock_get_spawner,
+        mock_get_store,
+        mock_spawn_wait,
+        mock_state_lock,
+        mock_build_prompt,
+        mock_read_draft,
+        mock_report,
+        mock_event_bus_emit,
+        mock_run_concurrent,
+    ):
+        """When the FAILED-marking block raises, a synthetic
+        PIPELINE_FAILED event must be published to the EventBus with
+        ``persisted=False`` and the chained error context.
+        """
+        from events import EventType
+        from routes.pipelines import _run_pipeline
+
+        pipeline = _make_running_pipeline()
+        mock_store, _ = _setup_mocks(
+            mock_report,
+            mock_read_draft,
+            mock_build_prompt,
+            mock_state_lock,
+            mock_spawn_wait,
+            mock_get_store,
+            mock_get_spawner,
+            mock_emit,
+            pipeline,
+        )
+
+        # Trigger the outer catch-all by raising a RuntimeError from
+        # ``_run_concurrent_phase`` — anything other than the
+        # ``(ContainerSpawnError, KubernetesSpawnError)`` the phase loop
+        # explicitly catches escapes to the outer ``except Exception``.
+        in_catchall = {"flag": False}
+
+        def concurrent_phase_raises(*args, **kwargs):
+            in_catchall["flag"] = True
+            raise RuntimeError("simulated outer-flow failure")
+
+        mock_run_concurrent.side_effect = concurrent_phase_raises
+
+        # Inner FAILED-marking block calls ``get_state_store`` fresh
+        # (line ~14615) and then ``store.load_pipeline``.  Make
+        # post-failure load_pipeline raise to simulate the racing
+        # ``_ensure_worktree`` that kills mark-FAILED in #2234.
+        def load_after_failure(*args, **kwargs):
+            if in_catchall["flag"]:
+                raise RuntimeError("simulated mark-failed crash")
+            return pipeline
+
+        mock_store.load_pipeline.side_effect = load_after_failure
+
+        with (
+            patch.dict(os.environ, {"EGG_HOST_REPO_MAP": '{"repo": "/host/repo"}'}, clear=False),
+            patch("pathlib.Path.exists", return_value=True),
+        ):
+            _run_pipeline("issue-42", Path("/repo"))
+
+        # The synthetic PIPELINE_FAILED event must have been published
+        # with persisted=False so wait-status callers can break out and
+        # surface the wedge.
+        synthetic_calls = [
+            c
+            for c in mock_event_bus_emit.call_args_list
+            if c.args and c.args[0] == EventType.PIPELINE_FAILED
+        ]
+        assert len(synthetic_calls) >= 1, (
+            f"Expected at least one PIPELINE_FAILED EventBus emit; "
+            f"got: {mock_event_bus_emit.call_args_list}"
+        )
+        unpersisted = [
+            c for c in synthetic_calls if c.kwargs.get("data", {}).get("persisted") is False
+        ]
+        assert len(unpersisted) == 1, (
+            f"Expected exactly one synthetic emit with persisted=False; got: {synthetic_calls}"
+        )
+        emit_call = unpersisted[0]
+        assert emit_call.args[1] == "issue-42"
+        data = emit_call.kwargs["data"]
+        assert data["status"] == PipelineStatus.FAILED.value
+        assert "simulated outer-flow failure" in data["original_error"]
+        assert "simulated mark-failed crash" in data["mark_error"]
+
+
 class TestFailurePathPushesWorktreeBranch:
     """Verify push_worktree_branch is called when pipeline.branch is set and
     the worktree path differs from the repo path."""

--- a/orchestrator/tests/test_pipeline_failure_path.py
+++ b/orchestrator/tests/test_pipeline_failure_path.py
@@ -26,6 +26,7 @@ sys.modules.setdefault("docker", MagicMock())
 sys.modules.setdefault("docker.errors", MagicMock())
 sys.modules.setdefault("docker.types", MagicMock())
 
+from events import EventType
 from gateway_client import GatewayError, PushResult
 from models import Pipeline, PipelinePhase, PipelineStatus
 
@@ -210,7 +211,6 @@ class TestZombiePipelineSyntheticEvent:
         PIPELINE_FAILED event must be published to the EventBus with
         ``persisted=False`` and the chained error context.
         """
-        from events import EventType
         from routes.pipelines import _run_pipeline
 
         pipeline = _make_running_pipeline()
@@ -238,10 +238,10 @@ class TestZombiePipelineSyntheticEvent:
 
         mock_run_concurrent.side_effect = concurrent_phase_raises
 
-        # Inner FAILED-marking block calls ``get_state_store`` fresh
-        # (line ~14615) and then ``store.load_pipeline``.  Make
-        # post-failure load_pipeline raise to simulate the racing
-        # ``_ensure_worktree`` that kills mark-FAILED in #2234.
+        # Inner FAILED-marking block calls ``get_state_store`` fresh and
+        # then ``store.load_pipeline``.  Make post-failure load_pipeline
+        # raise to simulate the racing ``_ensure_worktree`` that kills
+        # mark-FAILED in #2234.
         def load_after_failure(*args, **kwargs):
             if in_catchall["flag"]:
                 raise RuntimeError("simulated mark-failed crash")

--- a/orchestrator/tests/test_state_store.py
+++ b/orchestrator/tests/test_state_store.py
@@ -5,6 +5,7 @@ Note: Git operations are mocked since git init is not available in the sandbox.
 """
 
 import os
+import shutil
 import subprocess
 from unittest.mock import MagicMock, patch
 
@@ -1648,6 +1649,55 @@ class TestEnsureWorktreeIdempotency:
                     GitOperationError, match="Failed to remove stale state worktree"
                 ):
                     store._ensure_worktree()
+
+    def test_rmtree_enoent_treated_as_success(self, tmp_path):
+        """Regression for #2234.
+
+        `_ensure_worktree` runs from many threads concurrently (the state
+        store probe at ``state_store_probe.py``, every pipeline driver
+        thread polling via ``wait_for_decision``, the ``/api/v1/health``
+        probe, sibling pipelines on the same repo).  Between
+        ``wt.exists()`` (line ~244) and ``shutil.rmtree(wt)`` (line ~275),
+        a concurrent caller can rmtree the directory, leaving our rmtree
+        to raise ``FileNotFoundError``.  The directory being gone is
+        exactly the desired post-state — we must treat ENOENT as success
+        and continue to the recreate path, not raise ``GitOperationError``
+        and zombie the pipeline.
+        """
+        store, wt = self._make_store(tmp_path)
+        wt.mkdir()
+        (wt / ".git").write_text("gitdir: /broken\n")
+
+        # When our rmtree raises ENOENT, the recreate path needs wt gone
+        # on disk so the post-rmtree ``if wt.exists(): raise`` guard
+        # passes.  Simulate the racing deleter by removing wt for real.
+        # Bind the un-patched rmtree before patching so the helper's own
+        # delete does not recurse into the patched mock.
+        _real_rmtree = shutil.rmtree
+
+        def rmtree_after_concurrent_delete(path, *args, **kwargs):
+            _real_rmtree(path, ignore_errors=True)
+            raise FileNotFoundError(2, "No such file or directory", str(path))
+
+        with patch.object(
+            StateStore, "_run_git", side_effect=self._git_router(rev_parse_returncodes=[1, 1])
+        ) as mock_git:
+            with (
+                patch("state_store.time.sleep"),
+                patch("state_store.shutil.rmtree", side_effect=rmtree_after_concurrent_delete),
+            ):
+                # Must not raise — ENOENT means the racing caller already
+                # delivered the desired post-state.
+                result = store._ensure_worktree()
+
+        assert result == wt
+        # We must have continued through to recreate via `git worktree
+        # add`.  Without the ENOENT-as-success branch, this code path
+        # would have raised GitOperationError before reaching the add.
+        worktree_add_calls = [
+            c for c in mock_git.call_args_list if c.args[:2] == ("worktree", "add")
+        ]
+        assert len(worktree_add_calls) == 1
 
     def test_transient_rev_parse_failure_recovers_on_retry(self, tmp_path):
         """First rev-parse fails, second succeeds — worktree must be reused

--- a/orchestrator/tests/test_state_store.py
+++ b/orchestrator/tests/test_state_store.py
@@ -1656,13 +1656,14 @@ class TestEnsureWorktreeIdempotency:
         `_ensure_worktree` runs from many threads concurrently (the state
         store probe at ``state_store_probe.py``, every pipeline driver
         thread polling via ``wait_for_decision``, the ``/api/v1/health``
-        probe, sibling pipelines on the same repo).  Between
-        ``wt.exists()`` (line ~244) and ``shutil.rmtree(wt)`` (line ~275),
-        a concurrent caller can rmtree the directory, leaving our rmtree
-        to raise ``FileNotFoundError``.  The directory being gone is
-        exactly the desired post-state — we must treat ENOENT as success
-        and continue to the recreate path, not raise ``GitOperationError``
-        and zombie the pipeline.
+        probe, sibling pipelines on the same repo).  Between the
+        ``wt.exists()`` validity check and the ``shutil.rmtree(wt)`` that
+        the recreate path runs against the same path, a concurrent caller
+        can rmtree the directory, leaving our rmtree to raise
+        ``FileNotFoundError``.  The directory being gone is exactly the
+        desired post-state — we must treat ENOENT as success and continue
+        to the recreate path, not raise ``GitOperationError`` and zombie
+        the pipeline.
         """
         store, wt = self._make_store(tmp_path)
         wt.mkdir()

